### PR TITLE
feat: Add table to cross check user migration status

### PR DESCRIPTION
**Pull request by jrconlin**
This branch adds support to look in a `migration` table for a user's
`fxa_uid` record. If it is found, the server will return a 5xx error to
the client preventing sync operations.

NOTE: IT IS IMPORTANT THAT THE SERVER NOT RETURN A 2xx RESULT TO ANY
CLIENT IN MIGRATION STATUS UNLESS THE `meta`/`global` BSO IS RESET.

The `migration` table is created by the `user_migration` script, which
is part of `syncstorage-rs`.

### CODE REVIEW

File: syncstorage/storage/sql/dbconnect.py [hunk 1]
```diff
@@ -102,6 +102,21 @@ def __new__(cls, *args, **kwds):
...
+# This table contains the users in progress for migration. This table may
+# be updated by the external user migration script
+# A UID appearing in this table should cause the server to return a 5xx
+# response. It's important that the server NEVER return a 2xx response.
+
+migration = Table(
+    "migration",
+    metadata,
+    Column("fxa_uid", String(255), primary_key=True, nullable=False),
+    Column("started", BigInteger, nullable=False)
```
[the remark this answers is not in the document]
Comment (rfk) [reply syncstorage/storage/sql/dbconnect.py]: nit: it took me a long time to work out that you intended `started` to contain a timestamp, please consider calling this `started_at` or `start_time` or similar.

File: syncstorage/tests/test_sql.py [hunk 2]
```diff
@@ -197,3 +198,21 @@ def count_items():
...
+
+    def test_migrating(self):
+        migrating_id = str(uuid.uuid4())
+        with self.storage.dbconnector.connect() as connect:
+            sql = """
+                INSERT INTO migration
+                    (fxa_uid, started)
+                VALUES
+                    ('{}', {})
+                /* queryName=migrate */""".format(
```
Comment (rfk) [inline syncstorage/tests/test_sql.py]: nit: I know it's just a test, but you could use sql parameters here rather than string formatting:

```
connect.execute(sql, params={"fxa_uid": migrating_id, "started": int(time.time())})
```

File: syncstorage/views/decorators.py [hunk 3]
```diff
@@ -192,3 +196,19 @@ def with_collection_lock(viewfunc, request):
...
+    """Check if a given user is in a migration state, return 503.
+
+    NOTE: once a user migration has started DO NOT RETURN A 2xx
+    WITHOUT REMOVING THE meta/global RECORD!
+
+    """
+    storage = request.validated["storage"]
+    if storage.is_migrating(request.user):
+        raise HTTPServiceUnavailable(
+            headers={"Retry-After": MIGRATION_RETRY_AFTER}
```
Comment (rfk) [inline syncstorage/views/decorators.py]: Elsewhere in this codebase we use `str` to ensure that header values are strings before returning them, and I have vague recollections of errors that happen if you try to return an int in a header like this. According to #137 this works fine passing it as an integer, but you might like to use `str(MIGRATION_RETRY_AFTER)` here for consistency with other retry-header-setting code.

**Comment by rfk** [comment]
Please also consider the additional test from https://github.com/mozilla-services/server-syncstorage/pull/137

File: syncstorage/views/decorators.py [hunk 4]
```diff
@@ -28,6 +28,10 @@
 # How long the client should wait before retrying a conflicting write.
 RETRY_AFTER = 10
 
+# How long a client should wait to retry. It won't succeed, but should
+# return a 4xx error punting them to the new server.
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply syncstorage/views/decorators.py]: As I Understand, when a client gets a 4xx error, it needs to get a new token because it's been "reassigned". The migration script reaches into the tokenserver db as well to "assign" the user to spanner. I was originally thinking that I'd just wait until the client's token expires and the client naturally goes and checks for the new assignment, since that would potentially give the longest migration period.  That said, going to add a column to `migration` that is the `state`.

File: syncstorage/views/__init__.py [hunk 5]
```diff
@@ -179,6 +180,7 @@ def get_lbheartbeat(request):
 
 @info_timestamps.get(accept="application/json", renderer="sync-json",
                      acl=expired_token_acl)
+@check_migration
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply syncstorage/views/__init__.py]: This might be a styling nit for me. I'm not a huge fan of decorators to begin with because they secretly alter code flow and/or variables. Burying them inside a meta decorator that may or may not be applied doesn't really help, unless you keep what `default_decorators` does at top of mind at all times. 

That said, we'd need to do this check for any call the client might make which could return a 2xx result. I *could* add it to the `default_decorators` meta, and then add to the additional ones, but since this is kind of a semi-hack specific to the migration, I'd rather be very explicit about what's going on.

File: syncstorage/storage/sql/__init__.py [hunk 6]
```diff
@@ -397,6 +397,24 @@ def _find_items(self, session, user, collection, **params):
...
 
+    @with_session
+    def is_migrating(self, session, user):
+        """Returns if the user is migrating.
+
+        This should probably NOT be a cached call, since migration
+        may start at any time during a client sync.
+
+        """
+        userid = user.get("fxa_uid", user["uid"])
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply syncstorage/storage/sql/__init__.py]: I was mostly thinking about self hosters here, but you're right, they shoudn't be part of this at this time.

File: syncstorage/views/decorators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rfk) [reply syncstorage/views/decorators.py]: Gotcha. I think it would be OK to do as you suggest and just wait for the token to naturally expire, if you don't want the additional complexity (but in that case we should change the comment above to avoid confusion).

File: syncstorage/views/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rfk) [reply syncstorage/views/__init__.py]: I'm fine with leaving it separate if that fits best with your mental model 👍

File: syncstorage/views/decorators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rfk) [reply syncstorage/views/decorators.py]: Although do be aware that we can under some circumstances accept expired tokens for up to 2 hours after they expire, for reasons:

https://github.com/mozilla-services/server-syncstorage/blob/a7649edc1f02e529beb85a24c329f1c5ce108f4d/syncstorage/views/authentication.py#L18

I'm not sure we ever actually measured whether that was worth it in practice, we could consider removing it if we want users to hurry up and discover their migrated status.

File: syncstorage/tests/test_sql.py [hunk 7]
```diff
@@ -197,3 +198,21 @@ def count_items():
         self.assertEquals(count_items(), 5)
         self.assertEquals(len(self.storage.get_items(_USER, "col")["items"]),
                           5)
+
+    def test_migrating(self):
+        migrating_id = str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply syncstorage/tests/test_sql.py]: hrm, this is a bit tricky. The best way to test this is to add a test under `tests/functional/test_storage.py` which would add a record to a migration table and then test that the middleware layer returns a 503 when a matching fxa_uid record request is made. The problem is that there's a significant abstraction layer between the test architecture and the wsgi app such that I can't get to the wsgi storage from inside of the test framework. Since it's using `sqlite:////memory` that also complicates things. 

Another problem is that the `fxa_uid` is set to a random UUID in the `_authenticate()` function in `tests/function/support` so I can't just add a `testing` fxa_uid easily. 

I'm not really inclined to add a "set this user as migrating" type of call, even for testing, because that could be abused should it ever hit production. 

Any other test basically boils down to what's this test is doing, since the decorator is attached to all the relevant endpoints. 

I'll see if there's a way to get at least one full test, though.

[the remark this answers is not in the document]
Comment (jrconlin) [reply syncstorage/tests/test_sql.py]: Nevermind, I'm dumb and didn't read the patch below.

File: syncstorage/views/decorators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrconlin) [reply syncstorage/views/decorators.py]: Yeah, totally fine with changing the comment. The code is already weirdly complex because of the fact that it's loosely bound between two packages. QA might be a bit upset that they have to wait.

File: syncstorage/tests/test_sql.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rfk) [reply syncstorage/tests/test_sql.py]: lol; in this case  meant something a bit simpler though, basically taking the "assert that `is_migrating` ends up returning True" check from the bottom of this test, and copying it up here to the top as an "assert that `is_migrating` started out returning False" check.

File: syncstorage/storage/sql/__init__.py [hunk 8]
```diff
@@ -397,6 +397,24 @@ def _find_items(self, session, user, collection, **params):
...
+        try:
+            res = session.query_fetchone("MIGRATION_CHECK", {
+                "fxa_uid": userid
+            })
+            return res is not None
+        except Exception:
+            # if the db check fails, presume the user is not migrating.
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply syncstorage/storage/sql/__init__.py]: Adding some additional logic and passing the failure back as a 500

File: syncstorage/storage/sql/__init__.py [hunk 9]
```diff
@@ -397,6 +397,25 @@ def _find_items(self, session, user, collection, **params):
             "next_offset": next_offset,
         }
 
+    @with_session
+    def is_migrating(self, session, user):
+        """Returns if the user is migrating.
```
Comment (rfk) [inline syncstorage/storage/sql/__init__.py]: nit: either "Returns True if" or "Returns whether"

File: syncstorage/storage/sql/__init__.py [hunk 10]
```diff
@@ -397,6 +397,25 @@ def _find_items(self, session, user, collection, **params):
...
+
+        This should probably NOT be a cached call, since migration
```
Comment (rfk) [inline syncstorage/storage/sql/__init__.py]: In retrospect, it's not clear what you mean here by "be a cached call". Do you mean that it shouldn't be cached internally by this class, or not cached by callers of this class, or something else?

File: syncstorage/storage/sql/dbconnect.py [hunk 11]
```diff
@@ -102,6 +102,22 @@ def __new__(cls, *args, **kwds):
 )
 
 
+# Table listing migration state for users
+#
+# This table contains the users in progress for migration. This table may
+# be updated by the external user migration script
+# A UID appearing in this table should cause the server to return a 5xx
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: Nit: since most instances of "uid" or "userid" in this codebase refer to the numeric id allocated by tokenserver, it may be worth mentioning explicitly that this is the fxa uid in the comment here (which you do also mention by explicit name in the table declaration below 👍)

File: syncstorage/storage/sql/dbconnect.py [hunk 12]
```diff
@@ -363,6 +379,12 @@ def __init__(self, sqluri, create_tables=False, pool_size=100,
             collections.create(self.engine, checkfirst=True)
             user_collections.create(self.engine, checkfirst=True)
             batch_uploads.create(self.engine, checkfirst=True)
+            migration.create(self.engine, checkfirst=True)
+            # add this row for testing:
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: It seems unusual to have the production code add values specifically for testing, why do we need to do this here rather than e.g. in the test setup code?

File: syncstorage/tests/functional/support.py [hunk 13]
```diff
@@ -45,17 +45,17 @@ def new_do_request(req, *args, **kwds):
         orig_do_request = self.app.do_request
         self.app.do_request = new_do_request
 
-    def _authenticate(self):
+    def _authenticate(self, fxa_uid=None):
         # For basic testing, use a random uid and sign our own tokens.
-        # Subclasses might like to override this and use a live tokenserver.
+        # Subclasses might like to override this and use a live tokenserver.'
```
Comment (rfk) [inline syncstorage/tests/functional/support.py]: nit: spurious trailing `'`?

File: syncstorage/tests/functional/test_old_storage.py [hunk 14]
```diff
@@ -1,611 +0,0 @@
-# This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (rfk) [inline syncstorage/tests/functional/test_old_storage.py]: Was this file deleted on purpose? Removing it seems unrelated to the gist of the PR, but as always I could easily be missing something.

File: syncstorage/storage/sql/dbconnect.py [hunk 15]
```diff
@@ -102,6 +102,22 @@ def __new__(cls, *args, **kwds):
...
+# be updated by the external user migration script
+# A UID appearing in this table should cause the server to return a 5xx
+# response. It's important that the server NEVER return a 2xx response.
+
+migration = Table(
+    "migration",
+    metadata,
+    Column("fxa_uid", String(255), primary_key=True, nullable=False),
+    Column("started_at", BigInteger, nullable=False),
+    Column("state", String(32))  # unknown/NULL, in_progress, complete
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: @Micheletto, any concern about the space overhead of using a string here rather than e.g. a smallint enum?

File: syncstorage/tests/test_wsgiapp.py [hunk 16]
```diff
@@ -148,6 +139,31 @@ def new_do_request(req, *args, **kwds):
...
+        app = self._make_test_app(user)
+        req = self._make_signed_req(user["uid"], user)
+
+        # There's no public API for marking a user as migrated.
+        storage = get_storage(req)
+        with storage.dbconnector.connect() as connection:
+            connection.execute("""
+                INSERT INTO migration (fxa_uid, started_at)
+                VALUES (:fxa_uid, :started_at)
+                /* queryName=migrate */
```
Comment (rfk) [inline syncstorage/tests/test_wsgiapp.py]: nit: may want to add a `state` value here for completeness (I guess the test works because it defaults to `NULL` so it's not strictly necessary, but you use the string "in_progress" for a different test above).

File: syncstorage/storage/sql/dbconnect.py (hunk 12 above)
Comment (rfk) [reply syncstorage/storage/sql/dbconnect.py]: (Actually, having gotten through the rest of the PR, I don't see where this value is used in any tests - maybe it's a leftover from manual testing?)

File: syncstorage/storage/sql/__init__.py (hunk 10 above)
Comment (jrconlin) [reply syncstorage/storage/sql/__init__.py]: Honestly? Either. The problem right now is that migration is independent of the server, so a migration may begin during an active sync session. Since I don't believe that the migration server has the ability to alter memcache as well as the database, I don't believe that the server can properly cache this state. (I would be thrilled to be proven wrong here.)

I'm also not going to disagree that having the server check the table on call isn't fantastic. The alternate is possibly letting some corruption leak into the user record. I'm already trying my best to limit the potential by making the potential damage radius as small as possible. 

I guess it comes down to keeping the server safe or performant during migration.

File: syncstorage/storage/sql/dbconnect.py (hunk 12 above)
Comment (jrconlin) [reply syncstorage/storage/sql/dbconnect.py]: Yeah, it's left over from a previous test that was pulled. Thanks!

File: syncstorage/tests/functional/test_old_storage.py (hunk 14 above)
Comment (jrconlin) [reply syncstorage/tests/functional/test_old_storage.py]: No, it was by accident. I had renamed the file to aid in testing. It should be restored in the latest push.

File: syncstorage/storage/sql/dbconnect.py (hunk 15 above)
Comment (jrconlin) [reply syncstorage/storage/sql/dbconnect.py]: FWIW, I'm fine with whatever. I just need to make sure that this is coordinated with the other PRs related to migration.

**Comment by jrconlin** [comment]
Not going to argue about the potential performance impact. I'm really hoping that the migration state table is VERY temporary and we can disable it once migration has occurred. If you want, I *might* be able to put it behind an option flag since most stand-alone folk are probably not going to be migrating their data to spanner anytime soon.

**Comment by jrconlin** [comment]
waiting on @Micheletto's response regarding the state field.

File: syncstorage/storage/sql/__init__.py (hunk 10 above)
Comment (rfk) [reply syncstorage/storage/sql/__init__.py]: Right. Just to be clear, I'm not trying to argue for any sort of caching mechanism here - it just wasn't clear as a reader of the docstring exactly what it was trying to tell me to do or not do. I would support a more direct imperative like "Do not cache the result of this call, since...".

**Comment by rfk** [comment]
> If you want, I might be able to put it behind an option flag since most stand-alone folk
> are probably not going to be migrating their data to spanner anytime soon.

Self-hosters are probably least likely to be impacted by perf issues here, since they have so few users. It may be worth an option for Mozilla China though.

**Comment by jrconlin** [comment]
Agreed about China. Let me see what I can do about that.

File: syncstorage/storage/sql/dbconnect.py [hunk 17]
```diff
@@ -363,6 +379,7 @@ def __init__(self, sqluri, create_tables=False, pool_size=100,
             collections.create(self.engine, checkfirst=True)
             user_collections.create(self.engine, checkfirst=True)
             batch_uploads.create(self.engine, checkfirst=True)
+            migration.create(self.engine, checkfirst=True)
```
Comment (pjenvey) [inline syncstorage/storage/sql/dbconnect.py]: @Micheletto How have we done migrations (if any) in the past? Is `create_tables = true` (triggering this block) always specified in the config file?

Comment (jrconlin) [reply syncstorage/storage/sql/dbconnect.py]: From what I understand, migrations aren't always triggered, which is why i trap around any call to the migration table. The table can also be built by the migration script, which we will have a bit more control over.

**Comment by jrconlin** [comment]
Going to land this because @Micheletto is going to be out for a while. Can address any concerns of his later.

File: syncstorage/storage/sql/dbconnect.py (hunk 15 above)
Comment (erkolson) [reply syncstorage/storage/sql/dbconnect.py]: Bumping this

Comment (tublitzed) [reply syncstorage/storage/sql/dbconnect.py]: @jrconlin do we need to open a new PR here(and issue) to get this change in since it looks like this merged already?

Comment (jrconlin) [reply syncstorage/storage/sql/dbconnect.py]: Yes. see https://github.com/mozilla-services/syncstorage-rs/issues/428 &  https://github.com/mozilla-services/server-syncstorage/issues/143
